src/data/real/dedekind_cut.lean

```diff
@@ -0,0 +1,403 @@
+- TODO: finish constructing the full field
+
+## Tags
+
+dedekind, real, cut
+
+## References
+
+[Rudin, *Principles of Mathematical Analysis* (Chapter 1, Appendix)][rudin1976]
+
+-/
+
+section
+
+variables (Q : Type*) [has_lt Q]
+
+/-- *Dedekind cut*, defined for an arbitrary type with a less-than relation -/
+@[ext] structure dedekind_cut :=
+(carrier : set Q)
+(nonempty' : carrier.nonempty) -- Step 1, (I), first half
+(ne_top' : carrier ≠ ⊤) -- Step 1, (I), second half
+(is_lt_lower' : ∀ ⦃p q : Q⦄, q < p → p ∈ carrier → q ∈ carrier) -- Step 1, (II)
+(exists_gt' : ∀ ⦃p : Q⦄, p ∈ carrier → ∃ (r : Q) (hr : r ∈ carrier), p < r) -- Step 1, (III)
+
+variables {Q}
+
+namespace dedekind_cut
+
+instance : set_like (dedekind_cut Q) Q :=
+⟨dedekind_cut.carrier, λ p q h, by cases p; cases q; congr'⟩
+
+variables {α β γ : dedekind_cut Q} {p q r : Q}
+
+@[simp]
+lemma mem_carrier : p ∈ α.carrier ↔ p ∈ α := iff.rfl
+
+@[simp]
+lemma mem_coe : p ∈ (α : set Q) ↔ p ∈ α := iff.rfl
+
+@[simp]
+lemma mem_mk {s : set Q} (hn hne hl hg) : p ∈ mk s hn hne hl hg ↔ p ∈ s := iff.rfl
+
+variables (α)
+
+lemma nonempty : (α : set Q).nonempty := α.nonempty'
+lemma lt_top : (α : set Q) < ⊤ :=
+begin
+  split,
+  { simp },
+  have := α.ne_top',
+  contrapose! this,
+  ext,
+  simp only [mem_carrier, set.top_eq_univ, set.mem_univ, iff_true],
+  exact this (by simp)
+end
+lemma is_lt_lower ⦃p q : Q⦄ (h : q < p) (hp : p ∈ α) : q ∈ α := α.is_lt_lower' h hp
+lemma exists_gt ⦃p : Q⦄ (hp : p ∈ α) : ∃ (r : Q) (hr : r ∈ α), p < r := α.exists_gt' hp
+lemma exists_not_mem : ∃ (r : Q), r ∉ α :=
+begin
+  have := α.ne_top',
+  contrapose! this,
+  ext,
+  simpa using this _
+end
+
+lemma is_lower_set {Q : Type*} [partial_order Q] (α : dedekind_cut Q) : is_lower_set (α : set Q) :=
+begin
+  intros p q h hp,
+  rcases h.eq_or_lt with rfl|h,
+  { exact hp },
+  { exact α.is_lt_lower h hp }
+end
+
+lemma lt_of_mem_of_not_mem {Q : Type*} [linear_order Q] {α : dedekind_cut Q} {p q : Q}
+  (hp : p ∈ α) (hq : q ∉ α) : p < q :=
+begin
+  contrapose! hq,
+  exact α.is_lower_set hq hp
+end
+
+lemma not_mem_of_lt_of_not_mem {r s : Q} (h : r < s) (hr : r ∉ α) : s ∉ α :=
+mt (α.is_lt_lower h) hr
+
+noncomputable instance {Q : Type*} [linear_order Q] : linear_order (dedekind_cut Q) := -- Step 2
+{ le_total := λ α β, begin
+    by_cases h : α ≤ β,
+    { exact or.inl h },
+    { obtain ⟨p, hp, hnp⟩ : ∃ (p : Q) (hp : p ∈ α), p ∉ β,
+      { contrapose! h, exact h },
+      exact or.inr (λ q hq, α.is_lt_lower (lt_of_mem_of_not_mem hq hnp) hp) }
+  end,
+  decidable_le := classical.dec_rel (≤),
+  ..set_like.partial_order }
+
+-- Step 3
+/-- Dedekind cuts have the least-upper-bound property, where a Sup of a non-empty, bounded above
+set of cuts can be constructed. See `le_cSup` and `is_lub_cSup` for the resulting bound
+properties. -/
+def cSup (A : set (dedekind_cut Q)) (hA : A.nonempty) (hA' : bdd_above A) : dedekind_cut Q :=
```

Comment:
I would be happy to, but I cannot think of a proper definition for a total `Sup`. What would `Sup : set P -> P` send the empty set to, when the only information we have is that `has_lt P`?

Reply:
Ah yes, that's a problem. You might still want to provide said instance when `dedekind_cut Q` is `nonempty`.

Reply:
Right now, we don't have a `nonempty (dedekind_cut P)` instance defined anywhere. What are the conditions by which that holds?

Reply:
I don't think there's any simple necessary and sufficient set of typeclasses. `nontrivial` and `linear_order` suffices (since if `a` < `b`, the interval `Iio b` can be lifted to a Dedekind cut), though there's obviously examples where you don't have the latter. It might be best to just take `nontrivial (dedekind_cut P)` as a typeclass assumption itself.

Reply:
I think `nontrivial` + `linear_order` is fair game. Dedekind cuts are a weird bit (because they do not add the top and bottom) and as far as I'm aware they are only used to construct the reals. In other cases, the better way to do things is to take the Dedekind-MacNeille completion, which we already have in mathlib.